### PR TITLE
[4.0] Switcher Field

### PR DESCRIPTION
**Pull request by brianteeman**
Pull Request for Issue #20986 

### Summary of Changes
replace the custom element for the switcher field with pure css and accessible
it uses a legend to describe the fieldset and the label for the values as normal

### Testing Instructions
Requires npm i
- Component options
-  Module such as breadcrumbs
- featured switch on article edit 
#### Visual

bread

### Notes
1. The scss file isnt really scss - needs an expert eye to rewrite it
2. No RTL support - hopefully whoever fixes 1. will do that

### CODE REVIEW

**Comment by ghost** [comment]
I have tested this item :white_check_mark: successfully on 6f96df04c1b95d6bf03f4fae0745e2cdf704e2a2 using Module.

### System information
Setting | Value
-- | --
PHP Built On | Linux lamp122.cloudaccess.net 3.10.0-962.3.2.lve1.5.24.7.el6h.x86_64 nr.1 SMP Mon Dec 17 12:02:35 EST 2018 x86_64
Database Type | mysql
Database Version | 5.7.24-cll-lve
Database Collation | utf8_general_ci
Database Connection Collation | utf8_general_ci
PHP Version | 7.1.26
Web Server | Apache
WebServer to PHP Interface | cgi-fcgi
Joomla! Version | Latest Nightly Build Update from Joomla! 4.0.0-alpha8-dev Development [ Amani ] 9-March-2019 13:41 GMT
User Agent | Mozilla/5.0 (Macintosh; Intel Mac OS X 10.13; rv:66.0) Gecko/20100101 Firefox/66.0


<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/24463">issues.joomla.org/tracker/joomla-cms/24463</a>.</sub>

**Comment by Bakual** [comment]
As for the final location for the CSS file, I'd put it into https://github.com/joomla/joomla-cms/tree/4.0-dev/build/media_source/system/css/fields which ends up in /media/system/css/fields after building.
This way it can be overriden by the template if they choose so.

**Comment by brianteeman** [comment]
@bakual thanks that is what I was thinking too OR as scss to be compiled into the templates

**Comment by zwiastunsw** [comment]
1. Keyboard support OK
2. This is incomprehensible to me. In the context of the tested page on Joomla Backend, the screen reader (NVDA inc Chrome)
  - does not announce a switch label. Announces only states (Show, Hide),
  - treats all buttons as if it were one group of buttons and announces their status, e.g. 
> Show radio button checked 1 of 8
Hide radio button checked  4 od 8

instead:
> Show radio button checked 1 of 2
Hide radio button checked  2 od 2

When I copy the source code of a page and read it outside the context of the backend (as a separate document), the screen reader correctly reads the label of each switch and its status.   (Do not load any js then.)

**Firefox test**: switches do not receive keyboard focus at all
Can anyone confirm this?

**Comment by brianteeman** [comment]
Which exact page. Maybe there is a markup bug somewhere

**Comment by brianteeman** [comment]
I "think" I know what the problem is. It's not the code here. But could you please tell me the page you tested so I can confirm

**Comment by brianteeman** [comment]
@zwiastunsw I think it could be because the fieldsets are nested inside other fieldsets https://test-cases.tink.uk/nested-fieldsets/index.html

**Comment by brianteeman** [comment]
@zwiastunsw if that is the problem then we can use an alternative method instead of fieldset which is to use role=group

**Comment by zwiastunsw** [comment]
I tested  Site Modules > Breadcrumbs. 
Windows 10, Chrome v. 73.0.3683.86 (64 bit).

**Comment by brianteeman** [comment]
@zwiastunsw I am going to have to reach out to some external a11y experts on this. As you discovered the code works perfectly on a plain html page but not within joomla. I cant see anything in joomla that would break it.

**Comment by brianteeman** [comment]
@zwiastunsw @wilsonge I found the problem. It is NOT in the code here. The problem is in the tabs code. When you open a module for the first time ALL the sections are set with aria-hidden=true. The first tab should not be

image

**Comment by brianteeman** [comment]
@jeckodevelopment why did you update the branch? It doesn't have any purpose and forces my local branch to be out of date

**Comment by jeckodevelopment** [comment]
merge button was not available due to the branch being out of sync

**Comment by brianteeman** [comment]
@jeckodevelopment that is only relevant for the person who does the merge . it doesnt need to be done repeatedly. it has no impact on the ability to test the pr as long as there are no conflicts. All it does is waste your time pressing the button and my time updating my local branch

**Comment by wilsonge** [comment]
Let's chill :)

Is this now working from a screenreader perspective? @zwiastunsw

**Comment by brianteeman** [comment]
I'm chilled

**Comment by brianteeman** [comment]
@wilsonge Shall I go ahead and update this PR to use correct paths etc so that it can be merged?

**Comment by wilsonge** [comment]
Go for it

**Comment by brianteeman** [comment]
Updated code and original post with test instructions

**Comment by brianteeman** [comment]
The js tests are failing as expected as the js is no longer present. Should I remove the tests in this pr?

**Comment by brianteeman** [comment]
tests removed

File: tests/javascript/karma.conf.js [hunk 1]
```diff
@@ -1,87 +0,0 @@
...
-    // list of files / patterns to load in the browser
-    files: [
-      './**/fixtures/*.html',
-
-      // Include the webcomponents polyfills
-      { pattern: 'node_modules/@webcomponents/webcomponentsjs/custom-elements-es5-adapter.js', loaded: true, served: true, watch: false },
-      // Load web components polyfill
-      { pattern: 'media/vendor/webcomponentsjs/js/webcomponents-sd-ce-pf.js', loaded: true, served: true, watch: false },
-
-      // Load the files to test against
```
Comment (wilsonge) [inline tests/javascript/karma.conf.js:22]: I think we need to keep this file. But delete the 3 lines beneath here

File: build/media_source/system/joomla.asset.json [hunk 2]
```diff
@@ -49,6 +49,15 @@
         "media/system/js/showon.min.js"
       ]
     },
+    "switcher": {
+      "name": "switcher",
+      "dependencies": [
+        "core"
```
Comment (wilsonge) [inline build/media_source/system/joomla.asset.json]: why are we depending on `core` as a dependency here?

Comment (brianteeman) [reply build/media_source/system/joomla.asset.json]: It was my reading of the limited.docs.for this. Could be wrong

**Comment by wilsonge** [comment]
Thanks!

**Comment by brianteeman** [comment]
Woohoo

**Pull request merged**